src/braket/experimental/autoqasm/README.md

```diff
@@ -0,0 +1,137 @@
+which we have used as a foundation for this project.
+
+For a fully supported quantum developer experience,
+please continue to use the rest of the Amazon Braket Python SDK by following
+[these instructions](https://github.com/aws/amazon-braket-sdk-python#installing-the-amazon-braket-python-sdk).
+If you are interested in our active development efforts, and you are not
+afraid of a few bugs, please keep on reading!
+
+## Installation
+
+AutoQASM is an experimental module and is not yet part of the released Amazon Braket SDK.
+To use AutoQASM, you'll need to install directly from the `feature/autoqasm` branch:
+```
+git clone https://github.com/aws/amazon-braket-sdk-python.git
+cd amazon-braket-sdk-python
+git checkout feature/autoqasm
+pip install -e .
+```
+
+## Quick start
+
+In this section, we will show how to get started with AutoQASM. AutoQASM allows you to build
+quantum programs with a simplified syntax and run the programs on the service. It uses the circuit
+model programming paradigm that is also used in the Amazon Braket SDK.
+
+First, import the following modules and functions:
+```
+import braket.experimental.autoqasm as aq
+from braket.experimental.autoqasm.gates import h, cnot, measure
+```
+
+To create a quantum program using the AutoQASM experience, you decorate a function with `@aq.function`.
+This allows AutoQASM to hook into the program definition and generate an output format that is accepted
+by quantum devices.
+
+For instance, we can create a Bell state like so:
+```
+# A program that generates a maximally entangled state
+@aq.function
+def bell_state() -> None:
+    h(0)
+    cnot(0, 1)
+```
+
+You can view the output format, which is OpenQASM, by running `bell_state().to_ir()`.
+
+AutoQASM enables users to use more complicated program constructs with a compact and readable
+structure. We can demonstrate this with an active reset program:
+```
+# A program that actively resets a qubit back to the ground state
+@aq.function
+def reset(qubit: int, num_repeats: int) -> None:
+    for repeats in aq.range(num_repeats):
+        if measure(qubit):
+            x(qubit)
+```
+
+Now that the program takes inputs, you must pass those inputs when you call your function.
+In this case, the qubits are indexed by variable, rather than by integer literals, so we
+must additionally pass the `num_qubits` keyword argument to AutoQASM.
+```
+my_reset_program = reset(qubit=0, num_repeats=3, num_qubits=1)
+```
+
+AutoQASM can support nested subroutines and complex control flow. You can use the Python runtime
+and quantum runtime side-by-side. For the moment, we support only a few quantum operations such as
+`h`, `x`, `cnot`, and `measure`. There are rough edges at the moment, but we're actively smoothing
+them out!
+
+The Amazon Braket local simulator supports AutoQASM programs as input.
+Let's simulate the `my_reset_program` with the `mcm=True` argument,
+which enables simulation of mid-circuit measurements:
```

Comment:
Why do we need the mcm=True flag?

Reply:
It's an implementation detail of the particular fork of the LocalSimulator we use in this experimental branch. But thanks for raising this - we should just make this the default for AutoQASM programs. I've added a comment to this PR with a suggestion for where to do this. Then the user won't need to know about it.

Reply:
+1, to add some details, the main fork of the LocalSimulator is a circuit simulator, which performs all parsing/OpenQASM interpretation, then simulation, then measurement. For autoqasm, we need something more general that can intersperse classical and quantum computation, which I've implemented as a separate device, accessible by the `mcm` keyword.

MCM stands for mid-circuit measurement or Monte Carlo method, both ways that this simulator differs from the base version.